### PR TITLE
Up dev

### PR DESCRIPTION
**Pull request by turmary**
a esp_netconn_receive non-block timeout
new wifi/ap/netconn/dns apis

### CODE REVIEW

File: esp_at_lib/src/api/esp_netconn.c [hunk 1]
```diff
@@ -392,6 +392,45 @@ esp_netconn_connect(esp_netconn_p nc, const char* host, esp_port_t port) {
...
+ * \param[in]       nc: Netconn handle
+ * \param[in]       host: Pointer to host, such as domain name or IP address in string format
+ * \param[in]       port: Target port to use
+ * \param[in]       keep_alive: keep alive period seconds
+ * \param[in]       local_ip: local ip in connected command
+ * \return          \ref espOK if successfully connected, member of \ref espr_t otherwise
+ */
+espr_t
+esp_netconn_connect_ex(esp_netconn_p nc, const char* host, esp_port_t port, uint16_t keep_alive, const char* local_ip, esp_port_t local_port, uint8_t mode) {
+    esp_conn_start_t cs[1] = {0};
```
Comment (MaJerle) [inline esp_at_lib/src/api/esp_netconn.c]: Why did you decice for an array with 1 element instead of classic variable?

Comment (MaJerle) [reply esp_at_lib/src/api/esp_netconn.c]: This we need to re-think and create a esp_netconn_start_t (or similar) structure to add all parameters there. This is to ensure future features could be easily added w/o breaking compatibility. I will do this.

File: esp_at_lib/src/include/esp/esp_netconn.h [hunk 2]
```diff
@@ -75,6 +75,9 @@ int8_t          esp_netconn_getconnnum(esp_netconn_p nc);
 void            esp_netconn_set_receive_timeout(esp_netconn_p nc, uint32_t timeout);
 uint32_t        esp_netconn_get_receive_timeout(esp_netconn_p nc);
 
+espr_t          esp_netconn_connect_ex(esp_netconn_p nc, const char* host, esp_port_t port,
+                                       uint16_t keep_alive, const char* local_ip, esp_port_t local_port, uint8_t mode);
+
```
Comment (MaJerle) [inline esp_at_lib/src/include/esp/esp_netconn.h]: Good naming, I have to correct mistake of `esp_conn_startex` and rename it to `esp_conn_start_ex` instead.

File: esp_at_lib/src/api/esp_netconn.c [hunk 3]
```diff
@@ -673,6 +674,11 @@ esp_netconn_receive(esp_netconn_p nc, esp_pbuf_p* pbuf) {
 
     *pbuf = NULL;
 #if ESP_CFG_NETCONN_RECEIVE_TIMEOUT
+    if (nc->rcv_timeout == _RCV_NO_WAIT) {
+        if (!esp_sys_mbox_getnow(&nc->mbox_receive, (void **)pbuf)) {
+            return espTIMEOUT;
+        }
+    } else
```
Comment (MaJerle) [inline esp_at_lib/src/api/esp_netconn.c]: How do I set _RCV_NO_WAIT to rcv_timeout?

File: esp_at_lib/src/api/esp_netconn.c [hunk 4]
```diff
@@ -771,6 +816,16 @@ esp_netconn_get_receive_timeout(esp_netconn_p nc) {
...
 }
 
+/**
+ * \brief           Get netconn's conn handle
+ * \param[in]       nc: Netconn handle
+ * \return          esp conn handle.
+ */
+esp_conn_p
+esp_netconn_get_conn(esp_netconn_p nc) {
+    return nc->conn;
```
Comment (MaJerle) [inline esp_at_lib/src/api/esp_netconn.c]: Check for nc != NULL is necessary, not to dereference null pointer.

File: esp_at_lib/src/esp/esp.c [hunk 5]
```diff
@@ -256,6 +256,28 @@ esp_set_wifi_mode(esp_mode_t mode,
...
+ * \brief           Gets WiFi mode of either station only, access point only or both
+ *
+ * \param[in]       mode: point to space of Mode to get. This parameter can be a pointer of \ref esp_mode_t enumeration
+ * \param[in]       evt_fn: Callback function called when command has finished. Set to `NULL` when not used
+ * \param[in]       evt_arg: Custom argument for event callback function
+ * \param[in]       blocking: Status whether command should be blocking or not
+ * \return          \ref espOK on success, member of \ref espr_t enumeration otherwise
+ */
+espr_t
+esp_get_wifi_mode(esp_mode_t* mode,
```
Comment (MaJerle) [inline esp_at_lib/src/esp/esp.c]: Can `mode` pointer` be NULL?

File: esp_at_lib/src/esp/esp_ap.c [hunk 6]
```diff
@@ -182,6 +182,29 @@ esp_ap_configure(const char* ssid, const char* pwd, uint8_t ch, esp_ecn_t ecn, u
...
+ * \brief           Get configuration of Soft Access Point
+ *
+ * \note            Before you can get configuration access point, ESP device must be in AP mode. Check \ref esp_set_wifi_mode for more information
+ * \param[out]      ap_conf: soft access point configuration
+ * \param[in]       evt_fn: Callback function called when command has finished. Set to `NULL` when not used
+ * \param[in]       evt_arg: Custom argument for event callback function
+ * \param[in]       blocking: Status whether command should be blocking or not
+ * \return          \ref espOK on success, member of \ref espr_t enumeration otherwise
+ */
+espr_t esp_ap_get_conf(esp_ap_conf_t* ap_conf, const esp_api_cmd_evt_fn evt_fn, void* const evt_arg, const uint32_t blocking) {
```
Comment (MaJerle) [inline esp_at_lib/src/esp/esp_ap.c]: Inconsistent naming, `_conf` vs `_config`.

File: esp_at_lib/src/esp/esp_int.c [hunk 7]
```diff
@@ -695,6 +706,9 @@ espi_parse_received(esp_recv_t* rcv) {
 #endif /* ESP_CFG_HOSTNAME */
             } else if (CMD_IS_CUR(ESP_CMD_WIFI_CWDHCP_GET) && !strncmp(rcv->data, "+CWDHCP", 7)) {
                 espi_parse_cwdhcp(rcv->data);   /* Parse CWDHCP state */
+            } else if (CMD_IS_CUR(ESP_CMD_WIFI_CWMODE_GET) && !strncmp(rcv->data, "+CWMODE", 7)) {
+                const char* tmp = &rcv->data[8];/* Go to the number position */
+                *esp.msg->msg.wifi_mode.mode_get = (uint8_t)espi_parse_number(&tmp);
```
Comment (MaJerle) [inline esp_at_lib/src/esp/esp_int.c]: Mode cannot be NULL, so a check is necessary.

File: esp_at_lib/src/esp/esp_smart.c [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+ * \return          \ref espOK on success, member of \ref espr_t enumeration otherwise
+ */
+espr_t
+esp_smart_configure(uint8_t en,
+                    const esp_api_cmd_evt_fn evt_fn, void* const evt_arg, const uint32_t blocking) {
+    ESP_MSG_VAR_DEFINE(msg);
+
+    ESP_MSG_VAR_ALLOC(msg, blocking);
+    ESP_MSG_VAR_SET_EVT(msg, evt_fn, evt_arg);
+    ESP_MSG_VAR_REF(msg).cmd_def = en? ESP_CMD_WIFI_SMART_START: ESP_CMD_WIFI_SMART_STOP;
```
Comment (MaJerle) [inline esp_at_lib/src/esp/esp_smart.c]: `en? ESP_CMD_WIFI_SMART_START: ESP_CMD_WIFI_SMART_STOP` should be:
`en ? ESP_CMD_WIFI_SMART_START : ESP_CMD_WIFI_SMART_STOP`

File: esp_at_lib/src/include/esp/esp_ap.h [hunk 9]
```diff
@@ -56,6 +56,7 @@ espr_t      esp_ap_setip(const esp_ip_t* ip, const esp_ip_t* gw, const esp_ip_t*
 espr_t      esp_ap_getmac(esp_mac_t* mac, const esp_api_cmd_evt_fn evt_fn, void* const evt_arg, const uint32_t blocking);
 espr_t      esp_ap_setmac(const esp_mac_t* mac, const esp_api_cmd_evt_fn evt_fn, void* const evt_arg, const uint32_t blocking);
 
+espr_t      esp_ap_get_conf(esp_ap_conf_t* ap_conf, const esp_api_cmd_evt_fn evt_fn, void* const evt_arg, const uint32_t blocking);
```
Comment (MaJerle) [inline esp_at_lib/src/include/esp/esp_ap.h]: `get_config`

File: esp_at_lib/src/include/esp/esp_includes.h [hunk 10]
```diff
@@ -35,7 +35,8 @@
 #define ESP_HDR_INCLUDES_H
 
 #ifdef __cplusplus
-extern "C" {
+// don't nest extern "C"
+// extern "C" {
```
Comment (MaJerle) [inline esp_at_lib/src/include/esp/esp_includes.h]: Nesting of extern is OK, and just make sure it is always linked as C

**Comment by MaJerle** [comment]
Any feedback on this?

File: esp_at_lib/src/include/esp/esp_includes.h (hunk 10 above)
Comment (turmary) [reply esp_at_lib/src/include/esp/esp_includes.h]: Sorry for this late reply.
I know you method could worked well most cases.
A error condition for Example.
extern "C" {
#include "third_party.h"
}
will report error, if third_party.h is a C++ header, 
include class keywords which C can't recognized.

If we always let #include outside extern "C", and let every header owner make he's declarition as C linkage, we could avoid above problems.

Refer: http://effective-c.googlecode.com/svn-history/r67/wiki/EffectiveC.wiki

Comment (MaJerle) [reply esp_at_lib/src/include/esp/esp_includes.h]: turnary, you should NOT modify esp files.

The problem might be that extern C shall be after all includes. This I will review.

PS: Your link doesn't work.

**Pull request merged**